woocommerce/woocommerce
# @woocommerce/api: Complete SimpleProduct implementation

### All Submissions:


### Changes proposed in this Pull Request:

Currently the `AbstractProduct` and by proxy the `SimpleProduct` models lack a true implementation. This PR aims to add (almost) all of the missing properties, as well as the `list`, `read`, `update`, and `delete` repository methods.

### Architecture

As usual, I'd like to cover some of the meatier parts of this PR at a high-level to make review a little bit easier:

- `ModelTransformer` & `ModelTransformation`: Since one of the goals of this package is to abstract the model implementation from the REST API representation (for clarity and flexibility), it is necessary that some mechanism exist to transform between plain objects and model representations. I examined a few different options available on NPM but felt most of them were much heavier than we really needed and opted for something bespoke (but powerful!)

  The `ModelTransformer` class holds an array of `ModelTransformation` objects that are each designed to perform a specific kind of transformation to the properties. One noteworthy thing here is that the execution order of the transformations reverses on `toModel` so that things like key changes can happen first when digesting a plain object from the API.
- **Common REST Repository Functions**: After I started implementing the hooks for the `SimpleProduct` repository, it became clear that (almost?) all of our REST repositories will use the same basic logic. I took that as a sign that these functions should be made generic and refactored the existing repositories accordingly.

Closes #27871.

### How to test the changes in this Pull Request:

1. Place this snippet in `tests/e2e/api/simple-product-pr.test.ts`:

```typescript
import { HTTPClientFactory } from './http';
import { SimpleProduct } from './models';

// We have this random name generation function so that the test snippet works with a store that already has products!
function randomName(): string {
	const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

	let name = '';
	for ( let i = 64; i > 0; --i ) {
		name += characters[ Math.floor( Math.random() * characters.length ) ];
	}

	return name;
}

test( 'SimpleProduct Serialization', async () => {
	// We need a longer timeout or this test will fail.
	jest.setTimeout( 20000 );

	const httpClient = HTTPClientFactory.build( 'http://woocommerce.test' )
		.withBasicAuth( 'admin', 'password' )
		.create();

	const repository = SimpleProduct.restRepository( httpClient );

	const productName = randomName();

	const create = await repository.create( { name: productName, regularPrice: '199.99' } );
	expect( create ).toBeInstanceOf( SimpleProduct );
	expect( create.id ).toBeTruthy();
	expect( create.name ).toEqual( productName );
	expect( create.regularPrice ).toEqual( '199.99' );

	const list = await repository.list( { search: create.name } );
	expect( list ).toHaveLength( 1 );
	expect( list[ 0 ].id ).toEqual( create.id );
	expect( list[ 0 ].name ).toEqual( productName );
	expect( list[ 0 ].regularPrice ).toEqual( '199.99' );

	const read = await repository.read( create.id! );
	expect( read ).toBeInstanceOf( SimpleProduct );
	expect( read ).toMatchObject(
		{
			id: create.id,
			name: productName,
			price: '199.99',
			regularPrice: '199.99',
		},
	);

	const update = await repository.update( create.id!, { regularPrice: '9.99' } );
	expect( update ).toBeInstanceOf( SimpleProduct );
	expect( update.regularPrice ).toEqual( '9.99' );

	const d = await repository.delete( create.id! );
	expect( d ).toEqual( true );
} );
```
2. Adjust the `HTTPClientFactory` so that it will actually be able to make the requests to your local test site :)
3. Run `jest simple-product-pr` from the `tests/e2e/api` directory

### Other information:
